_posts/2022-05-12-faster-rand.md

```diff
@@ -0,0 +1,1096 @@
+[Measurement results] we see that the algorithm `exsp`
+that boasts sub-ns speed in C needs 17 ns in Erlang.
+
+32-bit Erlang is a sad story in this regard.  The bignum limit
+on such an Erlang system is so low, calculations would have
+to use 26-bit integers, that designing a PRNG
+not using bignums must be so small in period and size
+that it becomes too bad to be useful.
+The known trick `erlang:phash2(erlang:unique_integer(), Range)`
+is still fairly fast, but all `rand` generators work exactly the same
+as on a 64-bit system, hence operates on bignums so they are much slower.
+
+If your application needs a "random" integer for an non-critical
+purpose such as selecting a worker, choosing a route, etc,
+and performance is much more important than repeatability
+and statistical quality, what are then the options?
+
+
+
+Suggested solutions
+-------------------
+
+* [Use `rand` anyway]
+* [Write a BIF]
+* [Write a NIF]
+* [Use the system time]
+* [Hash a "unique" value]
+* [Write a simple PRNG]
+
+Reasoning and measurement results are in the following sections,
+but, in short:
+
+* Writing a NIF, we deemed, does not achieve a performance worth the effort.
+* Neither does writing a BIF.  But, ... a BIF (and a NIF, maybe) could
+  implement a combination of performance and quality that cannot
+  be achieved in any other way.  If a high demand on this combination
+  would emerge, we could reconsider this decision.
+* Using the system time is a bad idea.
+* `erlang:phash2(erlang:unique_integer(), Range)` has its use cases.
+* We have implemented a simple PRNG to fill the niche of non-critical
+  but very fast number generation: `mwc59`.
+
+
+### Use `rand` anyway
+
+Is `rand` slow, really?  Well, perhaps not considering what it does.
+
+In the [Measurement results] at the end of this text,
+it shows that generating a good quality random number using
+the `rand` module's default algorithm is done in 45 ns.
+
+Generating a number as fast as possible (`rand:mwc59/1`) can be done
+in less than 4 ns, but that algorithm has problems with the
+statistical quality.  See section [PRNG tests] and [Implementing a PRNG].
+
+Using a good quality algorithm instead (`rand:exsp_next/1`) takes 16 ns,
+if you can store the generator's state in a loop variable.
+
+If you can not store the generator state in a loop variable
+there will be more overhead, see section [Storing the state].
+
+Now, if you also need a number in an awkward range, as in not much smaller
+than the generator's size, you might have to implement a reject-and-resample
+loop, or even concatenate numbers.
+
+The overhead of code that has to implement this much of the features
+that the `rand` module already offers will easily approach
+its 26 ns overhead, so often there is no point in
+re-implementing this wheel...
+
+
+
+### Write a BIF
+
+There has been a discussion thread on Erlang Forums:
+[Looking for a faster RNG].  Triggered by this Andrew Bennett
+[potatosalad] wrote an experimental BIF.
```

Comment:
I would like to have links at the end of the document to not disrupt the Markdown text flow

Reply:
Yes, I figured as much. It is hard to make that type of suggestion in the github gui. My main point was that it would be nice with a link to the post with the bif.

Reply:
Fixed